### PR TITLE
Create FeeUnitFactory and functions to calculate tx fee rates

### LINKED ISSUE #1989: Add ability to calculate a `FeeUnit` from a `Transaction`
Currently, to calculate a fee rate for a transaction we have to do it manually, it would be nice and much safer if we had a way to take in a transaction and the funding amount and return a given `FeeUnit`.

Context: https://github.com/bitcoin-s/bitcoin-s/pull/1977#discussion_r485668105

### PR DESCRIPTION
**Pull request by benthecarman**
Closes #1989

### CODE REVIEW

File: core/src/main/scala/org/bitcoins/core/wallet/fee/FeeUnit.scala [hunk 1]
```diff
@@ -45,13 +70,38 @@ case class SatoshisPerByte(currencyUnit: CurrencyUnit) extends BitcoinFeeUnit {
     SatoshisPerKiloByte(currencyUnit.satoshis * Satoshis(1000))
   }
 
-  override def txSizeForCalc(tx: Transaction): Long = tx.byteSize
+  override def txSizeForCalc(tx: Transaction): Long =
+    SatoshisPerByte.txSizeForCalc(tx)
 
-  override def scaleFactor: Long = 1
+  override def scaleFactor: Long = SatoshisPerByte.scaleFactor
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/wallet/fee/FeeUnit.scala]: Seeing as this is the common pattern, I think you should have the top-level trait have an abstract def `FeeUnitFactory` and default impls like this ^ defined at the top level

File: core/src/main/scala/org/bitcoins/core/wallet/fee/FeeUnit.scala [hunk 2]
```diff
@@ -45,13 +70,38 @@ case class SatoshisPerByte(currencyUnit: CurrencyUnit) extends BitcoinFeeUnit {
...
+  override def txSizeForCalc(tx: Transaction): Long = tx.byteSize
+
+  override def calc(
+      inputAmount: CurrencyUnit,
+      tx: Transaction): SatoshisPerByte = {
+    val feePaid = inputAmount - tx.totalOutput
+    val feeRate = feePaid.satoshis.toLong / txSizeForCalc(tx).toDouble
+
+    fromLong(Math.round(feeRate))
+  }
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/wallet/fee/FeeUnit.scala]: why is this here? This seems to be equivalent to the thing it overrides?

Comment (benthecarman) [reply core/src/main/scala/org/bitcoins/core/wallet/fee/FeeUnit.scala]: Woops, missed in my cleanup

**Review by nkohen — APPROVED** [review]

File: core/src/main/scala/org/bitcoins/core/wallet/fee/FeeUnit.scala [hunk 3]
```diff
@@ -15,24 +15,53 @@ sealed abstract class FeeUnit {
...
+  /** Takes the given transaction returns a size that will be used for calculating the fee rate.
+    * This is generally the denominator in the unit, ie sats/byte
+    */
+  def txSizeForCalc(tx: Transaction): Long
+
+  /** Creates an instance T where the value given is the numerator of the fee unit
+    */
+  def fromLong(long: Long): T
+
+  def calc(inputAmount: CurrencyUnit, tx: Transaction): T = {
```
Comment (Christewart) [inline core/src/main/scala/org/bitcoins/core/wallet/fee/FeeUnit.scala]: What exactly is `inputAmount` here? A scala doc would be nice. You don't have an index if it's a specific input. If it's for all inputs we probably should rename to `totalInputAmount`

File: core/src/main/scala/org/bitcoins/core/wallet/fee/FeeUnit.scala [hunk 4]
```diff
@@ -15,24 +15,53 @@ sealed abstract class FeeUnit {
...
+    val feePaid = inputAmount - tx.totalOutput
+    val feeRate = feePaid.satoshis.toLong / txSizeForCalc(tx).toDouble
+
+    fromLong(Math.round(feeRate * scaleFactor))
```
Comment (Christewart) [inline core/src/main/scala/org/bitcoins/core/wallet/fee/FeeUnit.scala:67]: Should document the rounding style in the scaladoc as well

File: core/src/main/scala/org/bitcoins/core/wallet/fee/FeeUnit.scala [hunk 5]
```diff
@@ -93,12 +153,25 @@ case class SatoshisPerKiloByte(currencyUnit: CurrencyUnit)
...
+    SatoshisPerVirtualByte
 }
 
-object SatoshisPerVirtualByte {
+object SatoshisPerVirtualByte extends FeeUnitFactory[SatoshisPerVirtualByte] {
+
+  /** The coefficient the denominator in the unit is multiplied by,
+    * for example sats/kilobyte -> 1000
+    */
+  override def scaleFactor: Long = 1
```
Comment (Christewart) [inline core/src/main/scala/org/bitcoins/core/wallet/fee/FeeUnit.scala]: nit: `val`

File: core/src/main/scala/org/bitcoins/core/wallet/fee/FeeUnit.scala [hunk 6]
```diff
@@ -80,9 +122,27 @@ case class SatoshisPerKiloByte(currencyUnit: CurrencyUnit)
...
+  override def factory: FeeUnitFactory[SatoshisPerKiloByte] =
+    SatoshisPerKiloByte
+}
+
+object SatoshisPerKiloByte extends FeeUnitFactory[SatoshisPerKiloByte] {
 
+  /** The coefficient the denominator in the unit is multiplied by,
+    * for example sats/kilobyte -> 1000
+    */
   override def scaleFactor: Long = 1000
```
Comment (Christewart) [inline core/src/main/scala/org/bitcoins/core/wallet/fee/FeeUnit.scala]: nit `val`

**Review by Christewart — COMMENTED** [review]
Looks good overall!

**Review by Christewart — APPROVED** [review]

**Review by Christewart — APPROVED** [review]
